### DISCUSSION TITLE
Jupyter Notebook Link not Provided by Docker

### DISCUSSION THREAD

**Discussion by Fe-r-oz** [Xilinx/finn#854] [Q&A] [answered]
Dear Experts, 

I hope this question finds you well. I followed the following steps before facing a problem: The steps are:

- git clone https://github.com/Xilinx/finn/
- set up environment variables FINN_XILINX_PATH, FINN_XILINX_VERSION, PYNQ_IP, PYNQ_PORT, PYNQ_USERNAME 
- Install docker without root
- Run ./run-docker.sh quicktest
- Set up PYNQ board first-time setup cd /path/to/finn/ssh_keys => ssh-keygen =>ssh-copy-id -i id_rsa.pub <PYNQ_USERNAME>@<PYNQ_IP>=>ssh <PYNQ_USERNAME>@<PYNQ_IP>
- PYNQ_BOARD to export PYNQ_Board=Pynq-Z2
If this helps, Also, I am using Xilinx Vivado 2019.2 with Ubuntu Version and board that I am using is PYNQ-Z2.

After these steps, which all the "Quicktest" was passed, I ran **bash ./run-docker.sh notebook** and faced the following error as depicted by the snapshot. The problem I am facing is that**The jupyter notebook link is not provided by docker**

![Screenshot_select-area_20230713115459](https://github.com/Xilinx/finn/assets/93876775/242d760c-f149-4929-a252-839ce2806164)

As seen, the error says "board files folder mismatch, removing and downloading.
Kindly please provide me suggestions how to debug and solve this problem which is docker does not provide me access to jupyter notebook link.

**Thank you very much** for your assistance,
Sincerely
Feroz Ahmad

**Comment by auphelia** [comment, marked as answer]
Hi @Fe-r-oz ,

It looks like your docker installation was not successful. The error is thrown in [line 181](https://github.com/Xilinx/finn/blob/main/run-docker.sh#L181) which is the docker build command and line 246 which is the end of the file. 
The container doesn't give you a jupyter notebook link, because the script doesn't come to that point.

**Reply by Fe-r-oz** [reply]
Hi @auphelia, thanks for your reply.

These are the steps I took to try to solve the problem.

- Uninstall Docker from Ubuntu
- Reinstall Docker 
- Run  ./run-docker.sh quicktest
- Login to docker


**Error from daemon: Pull access denied for Xilinx/Finn, repository does not exist or require may docker login. Requested access to the resource is denied**
Snapshot is attached to represent the problem below
![Screenshot_select-area_20230716150251 (1)](https://github.com/Xilinx/finn/assets/93876775/a43fdcdc-7be1-4a50-867d-5bf3c99c6e01)

Kindly please suggest me the steps to solve the problem. I believe the problem is Finn-docker specific not ubuntu specific.

**Thank you very much** for your time and consideration.
Sincerely,
Feroz Ahmad

**Reply by auphelia** [reply]
Hi @Fe-r-oz ,
The error message `error: RPC failed; curl 18 transfer closed with outstanding read data remaining` looks like your internet connection closes while cloning the dependency repositories. You could try running the [fetch-repos.sh](https://github.com/Xilinx/finn/blob/main/fetch-repos.sh) script beforehand to ensure that the repos are already cloned before starting the docker script. But if your internet connection is not stable, you will most likely need to find a different workaround.

I think you have taken care of that, but please also make sure that you enabled non-root usage of docker for the FINN setup.

**Reply by Fe-r-oz** [reply]
Hi @auphelia, thank you very much for consistent support. Really grateful.

I successfully installed docker and passed the tests. But I face problem with running jupyter notebook with this command. 
- bash ./run-docker.sh notebook

Docker successful installation can be seen below:
![Screenshot_select-area_20230720104858](https://github.com/Xilinx/finn/assets/93876775/7112e7a8-5a69-4da2-8b05-280856a243ba)


The problem can be seen below:
![Screenshot_select-area_20230720111248](https://github.com/Xilinx/finn/assets/93876775/22e5d8b4-4d68-43c7-a04f-30e7bff6fc40)
Snapshot two of problem:
![Screenshot_select-area_20230720111350 (1)](https://github.com/Xilinx/finn/assets/93876775/fdf270ce-b32f-412a-a9bc-3481b27d5eff)

I set up these environment variables before running docker

- export FINN_XILINX_PATH=/home/Transcendence/Vivado/2019.1
- export VIVADO_PATH=/home/Transcendence/Vivado/2019.1
- export FINN_XILINX_VERSION=2019.1

I found this method from scrolling down the discussions. But "make sure that VIVADO_PATH is correctly mounted to docker container", I am not sure these commands complete that function.

Sincerely,
Feroz Ahmad

**Reply by fpjentzsch** [reply]
Hi,
the error relating to "attrs" might be fixed by a [recent commit](https://github.com/Xilinx/finn/pull/844/commits/391cd76ee3edb6e802d9b565a99993c775cc2194). Could you try to use the latest dev branch? You might run into problems due to your Xilinx tool version though, since we mainly support 2022.2 currently.

In the latest FINN version, you should only need to set the FINN_XILINX_PATH and FINN_XILINX_VERSION variables. The paths to Vivado and Vitis HLS will be assembled automatically and mounted into the Docker container.

Example if Vivado is installed in "/tools/Xilinx/Vivado/2022.2/...":
FINN_XILINX_PATH = /tools/Xilinx
FINN_XILINX_VERSION = 2022.2

**Reply by Fe-r-oz** [reply]
Hi @fpjentzsch, thank you for your reply.
I will follow the latest commit and will remove the error. 
I have observed several people (2) use 2019.1 with FINN 0.7v with no errors, maybe the support gets depreciated at every new version release.

In the meantime, could you please try to explain this error and suggest guidelines., It's a different error than "atrs" now. 
Please provide suggestions to remove it/
![Screenshot_select-area_20230724112821](https://github.com/Xilinx/finn/assets/93876775/75d19c91-32e2-44aa-ba26-1111be1f0e8c)

A different question: **Is possible and feasible to use FINN without Jupyter notebook support at all?** In order To bypass these jupyter notebook errors.

Best Regards
Feroz Ahmad

**Reply by fpjentzsch** [reply]
Hi,
it seems like you already have a Jupyter server running on your machine. You could tell FINN to use a different port via an environment variable.

Of course you can use FINN without starting up a Jupyter server. Just launch `./run-docker.sh ` to get an interactive shell where you can run Python scripts or use the "build_dataflow" or "build_custom" commands to invoke the build proces defined by a build.py script (see https://finn.readthedocs.io/en/latest/command_line.html).

**Comment by Tala22** [comment]
Hi 
i had the same problem , then i tried this code, then it is work properly :
docker run -it -p 8888:8888 jupyter/scipy-notebook:2023-08-19

**Comment by azer136** [comment]
Hi,

I had a similar problem that I solve by cloning the dev branch using: "git clone --branch dev https://github.com/Xilinx/finn.git". Using this solution, I could run the first two Jupyter Notebooks from /notebooks/end2end_example/cybersecurity (1-train-mlp-with-brevitas.ipynb and 2-import-into-finn-and-verify.ipynb) without problems (except with netron, but it's not my point). However, I could not execute correctly the last notebook (3-build-accelerator-with-finn.ipynb) and particularly the following line: build.build_dataflow_cfg(model_file, cfg_estimates).

I wonder if it is because my Vivado version's is 2019.2 and as @fpjentzsch already mentioned, the latest version of FINN could pose some problems with Vivado versions anterior to 2022.2. However, I couldn't find a way to launch the Docker of FINN with anterior versions (I tried v0.6, v0.7, v0.8, v0.8.1 and v0.9). All these versions seems to have problems with the library attrs, as mentioned previously.

Thus, my question is the following: "What version should I prefer to use with Vivado 2019.2 ? And, there is a way to solve the problem due to attrs library ?"

Thanks in advance for your reply.

PS: I do not provide errors report, because as I tried with several FINN version, I have different errors (globally all linked with the attrs library). Thus, I prefer to ask before what version of FINN I have to use with Vivado 2019.2, before sending the errors report corresponding to this version.
